libflux/flux-core/src/formatter/mod.rs

```diff
@@ -727,18 +729,21 @@ impl Formatter {
         if !n.body.is_empty() {
             self.indent()
         }
-
         let mut prev: i8 = -1;
+        let mut previous_location: i32 = -1;
         for (i, stmt) in n.body.iter().enumerate() {
             let cur = stmt.typ();
             self.write_rune(sep);
-
             if i != 0 {
-                // separate different statements with double newline or statements with comments
-                if cur != prev || starts_with_comment(Node::from_stmt(&stmt)) {
+                let current_location: i32 = stmt.base().location.start.line as i32;
+                //compare the line position of adjacent lines to preserve formatted double new lines
+                let line_gap = current_location - previous_location;
+                // separate different statements with double newline or statements with comments and preserve double newline format
+                if line_gap > 1 || cur != prev || starts_with_comment(Node::from_stmt(&stmt)) {
                     self.write_rune(sep);
                 }
             }
+            previous_location = stmt.base().location.end.line as i32;
```

Comment:
I would expect that the rules for formatting a file will be the same as the rules for formatting a block. Can you make them call the same common function, so this logic is not duplicated?

Reply:
Agree here as well, the line separation logic is duplicated and can be made common. In one of them, the first separator is only written in iterations 1 and up, and in the other it is written in all. That can be moved out of the common loop.

Other than that, looks good!